### DISCUSSION TITLE
Wait for component binding in `@xstate/svelte`

### DISCUSSION THREAD

**Discussion by alexamy** [statelyai/xstate#4091] [Q&A] [answered]
1. Write a machine that invokes some service on start
2. Use binded component exported methods in this service
3. Service is invoked before component is mounted

How to wait a svelte component to bind?  I know that this is inteded, but I don't know how to solve it properly. I've tried `onMount`, `tick`. Current ugly solution is additional state, which in transitioned manually after component is mounted. 
UPD: https://github.com/statelyai/xstate/discussions/4091#discussioncomment-6243508

Failed example (use devtools to see the error): https://stackblitz.com/edit/xstate-svelte-bind-component-order?file=src%2FApp.svelte
Working example (with additional initial state): https://stackblitz.com/edit/xstate-svelte-bind-component-order?file=src%2FAppWithInit.svelte

In this example you can add `async` / `tick`, and it's working, ~~but in large project this will not be sufficient~~ (needs investigation).

**Comment by Andarist** [comment]
I think that it's likely that we just shouldn't start the service in "render" but rather we should wait for `onMount` (or similar). Do you have any experience with this sort of thing? What are common expectations of Svelte users around this?

**Reply by alexamy** [reply]
I'm a new user in Svelte, so I dont know exactly how it should be done properly. But it makes sense, since Svelte has `onMount` helper.

Using `onMount` + `interpet` solves issue for demo project:
https://stackblitz.com/edit/xstate-svelte-bind-component-order-qrndkj?file=src%2FApp.svelte

I've tried to use this approach in current real project, but issue is still persist. So I will try to make complex example to demonstrate.

**Reply by alexamy** [reply]
There is a deferred render of referenced component in my project (data loading before ux is shown), so I need not only to wait mount, but also wait the defer to complete. After this changes `await tick()` is sufficient to solve my problem. With `onMount` + `interpret` I still need to use `await tick()`.
So maybe it is more a problem in my project structure than in Svelte wrapper for XState.

**Reply by Andarist** [reply, marked as answer]
This particular situation sounds like something to be solved on your side, in the app. I will still need to investigate if we need to do any changes in `@xstate/svelte` to cover simpler cases.

**Reply by alexamy** [reply]
Btw marking service with `async` in stackblitz example seems to solve this problem. In my project after moving init code in right place `await tick()` is sufficient to do the job. So it is not a big issue for me then. I've updated the part about large projects in the description.  Thanks for help!
